File: src/main/java/net/slipp/domain/qna/QuestionDto.java
package net.slipp.domain.qna;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import net.slipp.domain.fb.FacebookGroup;

import org.apache.commons.lang3.StringUtils;

import com.google.common.collect.Sets;

public class QuestionDto {
    private static final String FACEBOOK_GROUP_DELIMETER = "::";
    
    private Long questionId;
    
    private String title;

    private String contents;
    
    private String plainTags;
    
    private boolean connected = false;
    
    private String[] plainFacebookGroups;
    
    public QuestionDto() {
    }
    
    public QuestionDto(String title, String contents, String plainTags) {
        this(null, title, contents, plainTags);
    }
    
    public QuestionDto(Long questionId, String title, String contents, String plainTags) {
        this.questionId = questionId;
        this.title = title;
        this.contents = contents;
        this.plainTags = plainTags;
    }

    public Long getQuestionId() {
        return questionId;
    }

    public void setQuestionId(Long questionId) {
        this.questionId = questionId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getContents() {
        return contents;
    }

    public void setContents(String contents) {
        this.contents = contents;
    }

    public String getPlainTags() {
        return plainTags;
    }

    public void setPlainTags(String plainTags) {
        this.plainTags = plainTags;
    }
    
    public void setConnected(boolean connected) {
        this.connected = connected;
    }
    
    public boolean isConnected() {
        return this.connected;
    }
    
    public String[] getPlainFacebookGroups() {
        return plainFacebookGroups;
    }

    public void setPlainFacebookGroups(String[] plainFacebookGroups) {
        this.plainFacebookGroups = plainFacebookGroups;
    }
    
    public Set<FacebookGroup> getFacebookGroups() {
        return createFacebookGroups(this.plainFacebookGroups);
    }

    static Set<FacebookGroup> createFacebookGroups(String[] fbGroups) {
        if (fbGroups == null || fbGroups.length == 0) {
            return new HashSet<FacebookGroup>();
        }
        
        Set<FacebookGroup> groups = Sets.newHashSet();
        for (String each : fbGroups) {
            groups.add(createFacebookGroup(each));
        }
        return groups;
    }
    
    static FacebookGroup createFacebookGroup(String fbGroup) {
        if (StringUtils.isBlank(fbGroup)) {
            return new FacebookGroup.EmptyFacebookGroup();
        }
        String[] parsedGroups = fbGroup.split(FACEBOOK_GROUP_DELIMETER);
        if (parsedGroups.length != 2) {
            return new FacebookGroup.EmptyFacebookGroup();
        }
        return new FacebookGroup(parsedGroups[0], replaceSpaceToDash(parsedGroups[1]));
    }
    
    static String replaceSpaceToDash(String groupName) {
        if (StringUtils.isBlank(groupName)) {
            return "";
        }
        
        return groupName.replaceAll(" ", "-");
    }

    @Override
    public String toString() {
        return "QuestionDto [questionId=" + questionId + ", title=" + title + ", contents=" + contents + ", plainTags="
                + plainTags + ", connected=" + connected + ", plainFacebookGroups="
                + Arrays.toString(plainFacebookGroups) + "]";
    }
}

File: src/main/scala/net/slipp/domain/qna/QuestionDto.scala
package net.slipp.domain.qna

import java.lang.Long
import java.util.{HashSet, Set}

import com.google.common.collect.Sets
import net.slipp.domain.fb.{EmptyFacebookGroup, FacebookGroup}
import org.apache.commons.lang3.StringUtils

import scala.beans.BeanProperty

object QuestionDto {
  private val FACEBOOK_GROUP_DELIMETER: String = "::"

  private[qna] def createFacebookGroups(fbGroups: Array[String]): Set[FacebookGroup] = {
    if (fbGroups == null || fbGroups.length == 0) {
      return new HashSet[FacebookGroup]
    }
    val groups: Set[FacebookGroup] = Sets.newHashSet[FacebookGroup]
    for (each <- fbGroups) {
      groups.add(createFacebookGroup(each))
    }
    return groups
  }

  private[qna] def createFacebookGroup(fbGroup: String): FacebookGroup = {
    if (StringUtils.isBlank(fbGroup)) {
      return new EmptyFacebookGroup
    }
    val parsedGroups: Array[String] = fbGroup.split(FACEBOOK_GROUP_DELIMETER)
    if (parsedGroups.length != 2) {
      return new EmptyFacebookGroup
    }
    return new FacebookGroup(parsedGroups(0), replaceSpaceToDash(parsedGroups(1)))
  }

  private[qna] def replaceSpaceToDash(groupName: String): String = {
    if (StringUtils.isBlank(groupName)) {
      return ""
    }
    return groupName.replaceAll(" ", "-")
  }
}

class QuestionDto(qId: Long, t: String, c: String, pTags: String) {
  @BeanProperty
  var questionId: Long = qId

  @BeanProperty
  var title: String = t

  @BeanProperty
  var contents: String = c

  @BeanProperty
  var plainTags: String = pTags

  @BeanProperty
  var connected: Boolean = false

  @BeanProperty
  var plainFacebookGroups: Array[String] = null

  @BeanProperty
  var originalAnswerId: Long = _

  @BeanProperty
  var moveAnswers: Array[Long] = _

  def this(title: String, contents: String, plainTags: String) = this(null, title, contents, plainTags)

  def this(questionId: Long, originalAnswerId: Long, contents: String) = {
    this(questionId, null, contents, null)
    this.originalAnswerId = originalAnswerId
  }

  def this() = this(null, null, null, null)

  def getFacebookGroups: Set[FacebookGroup] = {
    return QuestionDto.createFacebookGroups(this.plainFacebookGroups)
  }

  override def toString = s"QuestionDto($questionId, $title, $contents, $plainTags, $connected, $plainFacebookGroups, $originalAnswerId, $moveAnswers)"
}
